File: frontend/src/utils/positionSeniority.ts
/**
 * Канонічний словник старшинства академічних посад (синхронізований з
 * `backend/.../common/PositionSeniority.java`).
 *
 * Менший ранг = вища посада. Невідома посада → UNKNOWN_RANK = 99.
 *
 * ВАЖЛИВО: порядок ключів має значення — `Object.entries(...)` повертає їх
 * у порядку вставки. Тому довші специфічні ключі («Старший науковий
 * співробітник») мають йти ПЕРЕД коротшими («Науковий співробітник»),
 * інакше у `rankOf` коротший ключ матчиться першим.
 */

export const UNKNOWN_POSITION_RANK = 99;

const SENIORITY: Record<string, number> = {
  'Начальник кафедри': 1,
  'Заступник начальника кафедри': 2,
  'Завідувач кафедри': 1,
  'Декан': 1,
  'Заступник декана': 2,
  'Професор': 3,
  // Має йти ПЕРЕД 'Науковий співробітник' — інакше будь-який "старший науковий"
  // буде матчитись на "науковий співробітник" і отримає ранг 7 замість 4.
  'Старший науковий співробітник': 4,
  'Доцент': 4,
  'Старший викладач': 5,
  'Викладач': 6,
  'Науковий співробітник': 7,
  'Асистент': 8,
};

/**
 * Повертає ранг посади за її назвою. Case-insensitive includes-match.
 * Якщо назва порожня або не матчиться жодним ключем — UNKNOWN_POSITION_RANK.
 */
export function rankOfPosition(positionTitle: string | null | undefined): number {
  if (!positionTitle) return UNKNOWN_POSITION_RANK;
  const lower = positionTitle.toLowerCase();
  for (const [key, rank] of Object.entries(SENIORITY)) {
    if (lower.includes(key.toLowerCase())) {
      return rank;
    }
  }
  return UNKNOWN_POSITION_RANK;
}

File: backend/src/main/java/ua/edu/teacherlicence/common/PositionSeniority.java
package ua.edu.teacherlicence.common;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Канонічний словник старшинства академічних посад.
 *
 * <p>Менший ранг = вища посада. Використовується:
 * <ul>
 *   <li>сортуванням таблиць (штат, кафедри);</li>
 *   <li>для вибору primary-посади викладача, що займає декілька штатних
 *       одиниць (Начальник кафедри > Доцент > Викладач ...).</li>
 * </ul>
 *
 * <p>Збігається з {@code frontend/src/utils/positionSeniority.ts}.
 *
 * <p><b>Порядок ключів важливий:</b> ітерація проводиться у порядку вставки
 * (LinkedHashMap). Тому довші ключі-префікси («Старший науковий співробітник»)
 * мають йти ПЕРЕД коротшими («Науковий співробітник»), інакше у методі
 * {@link #rankOf(String)} коротший ключ матчиться першим і повертає
 * неправильний ранг.
 */
public final class PositionSeniority {

    private PositionSeniority() {}

    /** Невідома посада — кладемо в кінець. */
    public static final int UNKNOWN_RANK = 99;

    /**
     * Словник «частина-назви-посади → ранг». Перевірка через
     * {@code title.toLowerCase().contains(key.toLowerCase())}.
     * Порядок ключів — від ширших/більш специфічних до загальних.
     */
    private static final Map<String, Integer> SENIORITY;
    static {
        SENIORITY = new LinkedHashMap<>();
        SENIORITY.put("Начальник кафедри", 1);
        SENIORITY.put("Заступник начальника кафедри", 2);
        SENIORITY.put("Завідувач кафедри", 1);
        SENIORITY.put("Декан", 1);
        SENIORITY.put("Заступник декана", 2);
        SENIORITY.put("Професор", 3);
        // ВАЖЛИВО: "Старший науковий співробітник" має йти ПЕРЕД "Науковий співробітник",
        // інакше "включає" матчиться на коротший ключ.
        SENIORITY.put("Старший науковий співробітник", 4);
        SENIORITY.put("Доцент", 4);
        SENIORITY.put("Старший викладач", 5);
        SENIORITY.put("Викладач", 6);
        SENIORITY.put("Науковий співробітник", 7);
        SENIORITY.put("Асистент", 8);
    }

    /**
     * Повертає ранг посади за її назвою. Case-insensitive includes-match.
     * Невідома посада → {@link #UNKNOWN_RANK}.
     */
    public static int rankOf(String positionTitle) {
        if (positionTitle == null || positionTitle.isBlank()) return UNKNOWN_RANK;
        String lower = positionTitle.toLowerCase();
        for (Map.Entry<String, Integer> entry : SENIORITY.entrySet()) {
            if (lower.contains(entry.getKey().toLowerCase())) {
                return entry.getValue();
            }
        }
        return UNKNOWN_RANK;
    }
}
